unity/Hello_World/Assets/Scripts/SpatialMappingCapture.cs

```diff
@@ -77,12 +94,89 @@ public class SpatialMappingCapture : MonoBehaviour, IMixedRealitySpatialAwarenes
     // Start is called before the first frame update
     protected void Start()
     {
+        Logger log = logger();
+
         // Create the image publisher and headset pose publisher
         ros = ROSConnection.GetOrCreateInstance();
         ros.RegisterPublisher<SpatialMeshMsg>(spatialMapTopicName);
 
         // Create the object detections 3d subscriber and register the callback
         ros.Subscribe<ObjectDetection3dSetMsg>(detectionsTopicName, DetectionCallback);
+
+        // Check that Open XR is available
+        var xrSettings = XRGeneralSettings.Instance;
+        var xrManager = xrSettings.Manager;
+        var xrLoader = xrManager.activeLoader;
+        var xrInput = xrLoader.GetLoadedSubsystem<XRInputSubsystem>();
+        if (xrInput != null)
+        {
+            // Set the origin tracking mode and recenter
+            xrInput.TrySetTrackingOriginMode(TrackingOriginModeFlags.Device);
+            xrInput.TryRecenter();
+#if ENABLE_WINMD_SUPPORT
+            unityCoordinateSystem = PerceptionInterop.GetSceneCoordinateSystem(UnityEngine.Pose.identity) as SpatialCoordinateSystem;
+#endif
+            xrInput.trackingOriginUpdated += new System.Action<UnityEngine.XR.XRInputSubsystem>(OriginUpdateHandler);
+        }
+
+        observer = CoreServices.GetSpatialAwarenessSystemDataProvider<IMixedRealitySpatialAwarenessObserver>();
+        if (observer == null)
+        {
+            log.LogInfo("Couldn't access Scene Understanding Observer!");
+            return;
+        }
+
+    }
+
+    private void OriginUpdateHandler(UnityEngine.XR.XRInputSubsystem s)
+    {
+#if ENABLE_WINMD_SUPPORT
+        unityCoordinateSystem = PerceptionInterop.GetSceneCoordinateSystem(UnityEngine.Pose.identity) as SpatialCoordinateSystem;
+#endif
+    }
+
+    protected void OnEnable()
+    {
+        CoreServices.SpatialAwarenessSystem.RegisterHandler<IMixedRealitySpatialAwarenessObservationHandler<SpatialAwarenessMeshObject>>(this);
+    }
+
+    protected void OnDisable()
+    {
+        CoreServices.SpatialAwarenessSystem.UnregisterHandler<IMixedRealitySpatialAwarenessObservationHandler<SpatialAwarenessMeshObject>>(this);
+    }
+
+    protected void OnDestroy()
+    {
+        CoreServices.SpatialAwarenessSystem.UnregisterHandler<IMixedRealitySpatialAwarenessObservationHandler<SpatialAwarenessMeshObject>>(this);
+    }
+
+    public void OnObservationAdded(MixedRealitySpatialAwarenessEventData<SpatialAwarenessMeshObject> eventData)
+    {
+        // Copy over the mesh data we need because we cannot access it outside of the main thread
+        Vector3[] vertices = eventData.SpatialObject.Filter.mesh.vertices;
+        int[] triangles = eventData.SpatialObject.Filter.mesh.triangles;
+        int id = eventData.SpatialObject.Id;
+        Thread tSendObject = new Thread(() => SendMeshObjectAddition(vertices,
```

Comment:
Are we sending this to a thread because otherwise this method would block and cause undue runtime slowdown?

Reply:
Yes, this callback is called from the main Unity thread, so the mesh sending pauses the frame rendering.